packages/component/src/providers/ActivityTree/ActivityTreeComposer.tsx

```diff
@@ -27,26 +34,35 @@ const ActivityTreeComposer = ({ children }: ActivityTreeComposerProps) => {
   const getKeyByActivity = useGetKeyByActivity();
   const activityKeys = useActivityKeys();
 
-  const activities = useMemo<readonly WebChatActivity[]>(() => {
-    const activities: WebChatActivity[] = [];
+  // Persistent Map to store activities by their keys
+  const activityMapRef = useRef<Readonly<Map<string, WebChatActivity>>>(
+    Object.freeze(new Map<string, WebChatActivity>())
+  );
 
-    if (!activityKeys) {
-      return rawActivities;
-    }
+  const activities =
+    useReduceActivities<readonly WebChatActivity[]>((prevActivities = [], activity) => {
+      if (!activityKeys) {
+        return rawActivities;
+      }
 
-    for (const activity of rawActivities) {
-      // If an activity has multiple revisions, display the latest revision only at the position of the first revision.
+      const activityKey = getKeyByActivity(activity);
 
       // "Activities with same key" means "multiple revisions of same activity."
-      const activitiesWithSameKey = getActivitiesByKey(getKeyByActivity(activity));
+      const activitiesWithSameKey = getActivitiesByKey(activityKey);
 
       // TODO: We may want to send all revisions of activity to the middleware so they can render UI to see previous revisions.
-      activitiesWithSameKey?.[0] === activity &&
-        activities.push(activitiesWithSameKey[activitiesWithSameKey.length - 1]);
-    }
+      if (activitiesWithSameKey?.[activitiesWithSameKey.length - 1] === activity) {
+        const activityMap = activityMapRef.current;
+
+        // Update or add the activity in the persistent Map
+        activityMap.set(activityKey, activity);
```

Comment:
If the `activity` is removed from the `rawActivities`, it will still keep in the `activityMapRef` forever and this will cause memory leak.